### PR TITLE
p2p/simulations: fix data race on swarm/network/simulations

### PR DESCRIPTION
**Pull request by justelad**
This PR addresses a data race caused by copying the node struct when creating a node event. This copy is done without synchronisation thus causing the data race warning.
Passing the underlying pointer instead eliminates the warning.

fixes https://github.com/ethersphere/go-ethereum/issues/1110

### CODE REVIEW

File: p2p/simulations/network.go [hunk 1]
```diff
@@ -168,7 +168,6 @@ func (net *Network) Start(id enode.ID) error {
 // snapshots
 func (net *Network) startWithSnapshots(id enode.ID, snapshots map[string][]byte) error {
 	net.lock.Lock()
-	defer net.lock.Unlock()
 
 	node := net.getNode(id)
 	if node == nil {
```
Comment (frncmx) [inline p2p/simulations/network.go]: I think if we get an error here, we will never unlock the mutex. This is true for all the error cases in this function.

File: p2p/simulations/network.go [hunk 2]
```diff
@@ -209,13 +210,14 @@ func (net *Network) watchPeerEvents(id enode.ID, events chan *p2p.PeerEvent, sub
 
 		// assume the node is now down
 		net.lock.Lock()
-		defer net.lock.Unlock()
 		node := net.getNode(id)
 		if node == nil {
```
Comment (frncmx) [inline p2p/simulations/network.go]: Same here. The lock might remain locked.

File: p2p/simulations/network.go (hunk 1 above)
Comment (justelad) [reply p2p/simulations/network.go]: good catch. fixed all of the corresponding error handlers

**Comment by frncmx** [comment]
I spent some time on this code. I was not happy with `Unlock()` scattered all over the place as that I find error prone. It's easy to miss a code path.

`defer` solves that problem elegantly. So I tried to extract a private method. However, that resulted in a different data-race. Basically, some tests just still read `n.Up` without a lock (example: `TestStartStopNode`).

I think the each `Node` should have its own lock for`Node.Up` and `Up` should be a private field accessed through setter/getters. Currently, not just that the whole `Network` blocks if a `Node` joins or leaves. But when `TestStartStopNode` calls `sim.StopNode` that calls `Network` where we lock to change a third type's (`Node`) internal state. Ehh.

----

Also, sometimes the test went into deadlock. I tried to investigate that, but instead I found: #1142

**Comment by frncmx** [comment]
@justelad Thanks for opening the other ticket.
Ship it! :+1:

**Comment by justelad** [comment]
@fjl @zsfelfoldi our merge is blocked due to the code owners file. could you please approve? we have a PR to amend the code owners file. thanks!

**Comment by frncmx** [comment]
fyi @nonsense as the code owner PR got merged this can be merged also.

**Pull request merged**